static/js/publisher/form/fileUpload.js

```diff
@@ -0,0 +1,82 @@
+import React from "react";
+
+import PropTypes from "prop-types";
+
+import { validateRestrictions } from "../../libs/fileValidation";
+
+class FileUpload extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.addFileHandler = this.addFileHandler.bind(this);
+    this.fileChangedHandler = this.fileChangedHandler.bind(this);
+    this.keyboardEventHandler = this.keyboardEventHandler.bind(this);
+  }
+
+  addFileHandler() {
+    const { restrictions, inputName, parent } = this.props;
+    const input = document.createElement("input");
+    input.type = "file";
+    if (restrictions.accept) {
+      input.accept = restrictions.accept.join(",");
+    }
+    input.name = inputName;
+    input.hidden = "hidden";
+    input.addEventListener("change", () => {
+      this.fileChangedHandler(input);
+    });
+
+    if (parent && parent.holder) {
+      parent.holder.appendChild(input);
+    } else {
+      this.holder.appendChild(input);
+    }
+
+    input.dispatchEvent(new Event("click", { bubbles: false }));
+  }
+
+  fileChangedHandler(input) {
+    const { fileChangedCallback, restrictions } = this.props;
+    const files = Array.from(input.files).map(file =>
+      validateRestrictions(file, restrictions)
+    );
+
+    Promise.all(files).then(values => {
+      fileChangedCallback(values);
+    });
+  }
+
+  keyboardEventHandler(e) {
+    if (e.key === "Enter") {
+      this.addFileHandler();
+    }
+  }
+
+  render() {
+    const { className, children } = this.props;
+    return (
+      <div
+        ref={el => (this.holder = el)}
+        className={className}
+        onClick={this.addFileHandler}
+        onKeyDown={this.keyboardEventHandler}
+        tabIndex={0}
+      >
+        {children}
+      </div>
+    );
+  }
+}
+
+FileUpload.propTypes = {
+  className: PropTypes.string,
+  children: PropTypes.node,
+  inputName: PropTypes.string,
+  parent: PropTypes.node,
```

Comment:
Having a second look at that it seems that we are passing a component here, not a node (but I'm not sure how they correlate to each other exactly in React).
Looking at the code it seems that we are interested in `parent.holder` so maybe a shape expecting `holder` node would be better?

Reply:
BTW, that's quite a strange pattern for React anyway, to pass a component and do something with it.
Is it necessary? Can't this component just render inputs in itself? Is there a case when input really has to be put elsewhere in the DOM?

Reply:
Yeah, the issue is that the `parent.holder` node isn't defined when `FileUpload` get's rendered (as it's a child of the parent). While typing that I did think maybe if `this.holder` was defined in the constructor of the pare, when the element has been rendered the reference will be updated.... Leave it with me, I want to try something :)